### DISCUSSION TITLE
Where does `remark-math` declare its HTML?

### DISCUSSION THREAD

**Discussion by benrbray** [remarkjs/remark#1338] [Q&A]
## Overview

* I am attempting to extend my [`remark-cite`](https://github.com/benrbray/remark-cite) package with a `rehype-cite` plugin for generating HTML
* I'm using `remark-math` and `rehype-katex` as a reference for how to accomplish this
* I can't figure out where the `remark-math` plugin passes the `mathHtml` extension to unified

## Details

I know that `rehype-katex` works by looking for `math-inline` and `math-display` elements in the `hast` tree:

https://github.com/remarkjs/remark-math/blob/dd3075a597d243671bf2bbb46400e3e0f3577078/packages/rehype-katex/lib/index.js#L47-L56

The `math-inline` and `math-display` classes are assigned by the `mathHtml` plugin exported by `micromark-extension-math`:

https://github.com/micromark/micromark-extension-math/blob/6fd7df3338fee9a2490cd32bee54d71c0af48581/dev/lib/html.js#L33-L46

If we have a processor defined as follows:

```typescript
import remarkMath from "remark-math";

const processor = unified()
  .use(remarkParse)
  .use(remarkMath)
  .use(remarkRehype)

const markdown = "look it's $math$!";
const mdast = processor.parse(markdown);
const hast = processor.runSync(mdast);
```

The `hast` tree will contain a `math-inline` node:

![image](https://github.com/remarkjs/remark/assets/221963/665037e9-72d1-47c6-8d35-1de67b9bbace)

However, the [`remark-math` repository](https://github.com/remarkjs/remark-math/tree/main) does not include any reference to the `mathHtml` plugin exported by `micromark-extension-math`.

https://github.com/search?q=repo%3Aremarkjs%2Fremark-math%20mathHtml&type=code

(see github search results above -- but also dug through the code manually)

## Questions

* how does `remark-math` know how to generate HTML for math nodes if `mathHtml` is never imported?
* if I have a remark plugin that defines a new node type, what's the best way to allow that node type to be detected & processed by a rehype plugin?

**Comment by ChristianMurphy** [comment]
Hey @benrbray 👋 
Went dealing with known or new node types that you want specific HTML output, leverage `.data.hName`, `.data.hProperties`, and `.data.hChildren` https://github.com/syntax-tree/mdast-util-to-hast?tab=readme-ov-file#fields-on-nodes

That is also how `remark-math` works https://github.com/syntax-tree/mdast-util-math/blob/23647664692ae59be25896d0099ba3f7157d6a76/lib/index.js#L53-L68

**Reply by wooorm** [reply]
Yup, as Christian mentions.

I’d recommend doing as little as possible in `h*` and such.
You have to think about: what if someone uses my syntax in plain markdown: they parse and format and serialize markdown. Will it work well? Will my markdown/remark/mdast integrations work well?
And also: what if someone writes HTML manually? Will my html/rehype/hast integrations work well for them, too?
